rfcs/0052-dynamic-ids.md

```diff
@@ -0,0 +1,121 @@
+
+In addition, a central list of ids is annoying to maintain and leads to merge conflicts.
+
+# Detailed design
+[design]: #detailed-design
+
+## Documentation updates
+
+New best practices for declaring users are to be documented. This includes sections like the following
+
+### Using `DynamicUser` for declaring users
+
+If the service is fit for `DynamicUser`, this is the preferred solution, as it doesn't require any persistent ids. However this only works well for services whose state is self-contained. As soon as the service's data is needed somewhere else, this approach can fall flat. `DynamicUser` also enables a bunch of options restricting the service (see `man systemd.exec` for details), some of which might need to be disabled in order for it to work. Note that `DynamicUser` works even with large amounts of files, since it almost never has to change the underlying uid which would require a `chown` of all files.
+
+An example:
+
+```nix
+{
+  systemd.services.myservice.serviceConfig.DynamicUser = true;
+}
+```
+
+### Using ids dynamically allocated by NixOS
+
+
+By not setting `users.users.<name?>.uid`/`users.groups.<name?>.gid`, NixOS will dynamically allocate ids. For users, `users.users.<name?>.isSystemUser = true` should be set as well such that only ids below 1000 are used. These ids are persistent over the lifetime of a NixOS system, even when services are disabled and enabled again. The generated mapping from names to ids is stored in `/var/lib/nixos/uid-map`/`/var/lib/nixos/gid-map`, so if this directory is backed up, the mappings will persist too when restoring.
+
+An example:
+
+```nix
+{
+  users = {
+    users.myservice = {
+      description = "My service user";
+      group = "myservice";
+      isSystemUser = true;
+    };
+    groups.myservice = {};
+  };
+}
+```
+
+### Ensuring correct directory permissions for services
+
+Services must be able to access their directories.
+- The easiest way to achieve this is to use `systemd.services.<name?>.serviceConfig.StateDirectory = "myservice"`, which ensures that `/var/lib/myservice` belongs to the services user. See `man systemd.exec` for info on this and the related directives `CacheDirectory`, `LogsDirectory` and `ConfigurationDirectory`.
+- `systemd.tmpfiles.rules = [ "Z '/var/lib/myservice' - myuser mygroup - -" ]` can also be used, with the disadvantage that it will only run at system activation and not when the service starts. It also recursively fixes the permissions every time, meaning it can lead to considerable slowdown with many files.
+- An alternative is to assign `serviceConfig.ExecStartPre = "+${pkgs.writeScript "myservice-prestart" "..."}"` with a script to fix the permissions, where the `+` makes the script run with full root permissions as documented in `man systemd.service`.
+
+## Enlarge the reserved range of system users/groups
```

Comment:
In light of @asymmetric 's [comment]( https://github.com/NixOS/nixpkgs/pull/65698#issuecomment-532835429) about how systemd ignores `login.defs` this section might need to be slightly tweaked, or maybe the PR needs to be fixed to have systemd and login.defs depend on the same nix definition.

Reply:
See https://github.com/NixOS/nixpkgs/pull/65698#issuecomment-537918789, I'll mention here that people should see that PR for implementation details

Reply:
So here `SYS_UID_MIN` and such would be representative of whatever needs to be changed to get the change working (and really, `SYS_UID_MIN` should be the way to configure this, I think it's a bug that it doesn't work).

Reply:
Can we mark this as resolved?